File: cpp/696_count_binary_substrings/count_binary_substrings.cpp
#include <cassert>
#include <string>
#include <tuple>
#include <vector>
using namespace std;

class Solution {
 public:
  int countBinarySubstrings(string s) {
    int ptr = 0, n = s.size(), last = 0, ans = 0;
    while (ptr < n) {
      char c = s[ptr];
      int count = 0;
      while (ptr < n && s[ptr] == c) {
        ++ptr;
        ++count;
      }
      ans += min(count, last);
      last = count;
    }
    return ans;
  }
};

int main() {
  vector<tuple<string, int>> tests{
      {"00110011", 6},
      {"10101", 4},
  };

  for (auto& [s, ans] : tests) {
    assert(Solution().countBinarySubstrings(s) == ans);
  }
  return 0;
}

File: golang/696_count_binary_substrings/count_binary_substrings.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func countBinarySubstrings(s string) int {
	var ptr, last, ans int
	n := len(s)
	for ptr < n {
		c := s[ptr]
		count := 0
		for ptr < n && s[ptr] == c {
			ptr++
			count++
		}
		ans += min(count, last)
		last = count
	}

	return ans
}

func main() {
	tests := []struct {
		s   string
		ans int
	}{
		{"00110011", 6},
		{"10101", 4},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, countBinarySubstrings(test.s), test.ans, index)
	}
}
